webpanel/login.go

```diff
@@ -1,46 +1,98 @@
+	"crypto/rand"
+	"encoding/base64"
+	"io"
 	"net/http"
 	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/golang-jwt/jwt/v5"
+	"golang.org/x/oauth2"
 )
 
-const QueryUser = `SELECT password FROM panel_user WHERE username = $1`
 
-func (w *WebPanel) LoginPage(c *gin.Context) {
-	c.HTML(http.StatusOK, "login.html", nil)
+
+func randString(nByte int) (string, error) {
+	b := make([]byte, nByte)
+	if _, err := io.ReadFull(rand.Reader, b); err != nil {
+		return "", err
+	}
+	return base64.RawURLEncoding.EncodeToString(b), nil
 }
 
-func (w *WebPanel) Login(c *gin.Context) {
-	username := c.PostForm("username")
-	password := c.PostForm("password")
+func setCallbackCookie(w http.ResponseWriter, r *http.Request, name, value string) {
+	c := &http.Cookie{
+		Name:     name,
+		Value:    value,
+		MaxAge:   int(time.Hour.Seconds()),
+		Secure:   r.TLS != nil,
+		HttpOnly: true,
+	}
+	http.SetCookie(w, c)
+}
+	
+func (w *WebPanel) StartPanelHandler(c *gin.Context) {
+	state, err := randString(16)
+	if err != nil {
+		c.HTML(http.StatusInternalServerError, "error.html", gin.H{
+			"Error": err.Error(),
+		})
+		return
+	}
+	
+	setCallbackCookie(c.Writer, c.Request, "state", state)
+	
+	http.Redirect(c.Writer, c.Request, w.AuthConfig.OAuth2Config.AuthCodeURL(state), http.StatusFound)
+}
 
-	var expectedHash string
-	err := w.Pool.QueryRow(w.Ctx, QueryUser, username).Scan(&expectedHash)
+func (w *WebPanel) FinishPanelHandler(c *gin.Context) {
+	//TODO: implement response handler for OIDC
+	state, err := c.Request.Cookie("state")
 	if err != nil {
-		c.HTML(http.StatusUnauthorized, "login.html", gin.H{
-			"Error": "Database error",
+		c.HTML(http.StatusBadRequest, "error.html", gin.H{
+			"Error": "State cookie not found",
 		})
 		return
 	}
 
-	hashBytes := sha512.Sum512(append(w.Salt, []byte(password)...))
-	fmt.Println(hex.EncodeToString(hashBytes[:]))
-	if hex.EncodeToString(hashBytes[:]) != expectedHash {
-		c.HTML(http.StatusUnauthorized, "login.html", gin.H{
-			"Error": "Authentication failed",
+	if c.Request.URL.Query().Get("state") != state.Value {
+		c.HTML(http.StatusBadRequest, "error.html", gin.H{
+			"Error": "State did not match",
 		})
 		return
 	}
 
-	// Now that we know the user is valid, generate a JWT.
+	oauth2Token, err := w.AuthConfig.OAuth2Config.Exchange(c, c.Request.URL.Query().Get("code"))
+	if err != nil {
+		c.HTML(http.StatusBadRequest, "error.html", gin.H{
+			"Error": err.Error(),
+		})
+		return
+	}
+
+	userInfo, err := w.AuthConfig.Provider.UserInfo(c, oauth2.StaticTokenSource(oauth2Token))
+		if err != nil {
+			http.Error(c.Writer, "Failed to get userinfo: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+	/* resp := struct {
+			OAuth2Token *oauth2.Token
+		}{oauth2Token}
+		data, err := json.MarshalIndent(resp, "", "    ")
+		if err != nil {
+			c.HTML(http.StatusInternalServerError, "error.html", gin.H{
+				"Error": err.Error(),
+			})
+			return
+		}
+		c.Writer.Write(data) */
```

Comment:
Figure out if you are going to use this.

Reply:
This part is a leftover of the non-OIDC login. I'll remove it